### PR TITLE
Add memory efficient attention from xformers

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
- Add the memory efficient attention kernels from xformers into core.
- Didn't wire up the backward to the forward in derivatives.yaml yet
- Put these files next to the flash_attention files. 
- Build system is using the same `USE_FLASH_ATTENTION` flag which Christian is working on removing.

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -13170,6 +13170,11 @@
   dispatch:
     CUDA: flash_scaled_dot_product_attention
 
+- func: _efficient_attention(Tensor query, Tensor key, Tensor value, bool compute_log_sumexp=False, Tensor? attn_bias=None, float p=0.0, bool causal=False) -> (Tensor, Tensor)
```
Comment (drisspg) [inline aten/src/ATen/native/native_functions.yaml:13173]: native functions for some reason doesn't let you return `int` inside of a tuple.

**Comment by cpuhrsch** [comment]
@drisspg - can you also add measurements for the CUDA context size?

**Comment by drisspg** [comment]
On master build:
```
 nvidia-smi --query-gpu=utilization.memory,memory.total,memory.free,memory.used --format=csv -i 0
utilization.memory [%], memory.total [MiB], memory.free [MiB], memory.used [MiB]
0 %, 40960 MiB, 39538 MiB, 815 MiB
```

```
❯  nvidia-smi --query-gpu=utilization.memory,memory.total,memory.free,memory.used --format=csv -i 0
utilization.memory [%], memory.total [MiB], memory.free [MiB], memory.used [MiB]
0 %, 40960 MiB, 39528 MiB, 825 MiB
```

10mb jump but I also think this is including both the xformers and flash_attention code

**Comment by drisspg** [comment]
Closing because changes will be landed in: #85281

**Pull request closed without merging**